### PR TITLE
Add repeat interleave op

### PR DESCRIPTION
**Pull request by BBuf**
- add repeat interleave op.

docstring：

image

image

### CODE REVIEW

File: oneflow/core/functional/impl/array_functor.cpp [hunk 1]
```diff
@@ -2649,7 +2651,7 @@ class IndexSelectFunctor {
 
     DimVector index_broad_cast(input_num_axes);
     for (int i = 0; i < input_num_axes; i++) { index_broad_cast[i] = input->shape()->At(i); }
-    index_broad_cast[new_dim] = 1;
+    if (index_broad_cast[new_dim] != 0) { index_broad_cast[new_dim] = 1; }
```
Comment (BBuf) [inline oneflow/core/functional/impl/array_functor.cpp:2654]: For 0-size tensor

File: python/oneflow/framework/docstr/repeat_interleave.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+"""
+import oneflow
+from oneflow.framework.docstr.utils import add_docstr
+
+add_docstr(
+    oneflow.repeat_interleave,
+    """
+    repeat_interleave(input, repeats, dim=None, *, output_size=None) -> Tensor
+
+    The documentation is referenced from: https://pytorch.org/docs/stable/generated/torch.repeat_interleave.html#torch.repeat_interleave.html
```
Comment (zhongshsh) [inline python/oneflow/framework/docstr/repeat_interleave.py:24]: The documentation is referenced from: https://pytorch.org/docs/stable/generated/torch.repeat_interleave.html

File: oneflow/user/kernels/repeat_interleave_kernel.cpp [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+};
+
+#define REGISTER_REPEAT_INTER_LEAVE_KERNEL(dtype)                     \
+  REGISTER_USER_KERNEL("repeat_interleave")                           \
+      .SetCreateFn<CpuRepeatInterLeaveKernel<dtype>>()                \
+      .SetIsMatchedHob((user_op::HobDeviceType() == DeviceType::kCPU) \
+                       && (user_op::HobDataType("in", 0) == GetDataType<dtype>::value))
+
+REGISTER_REPEAT_INTER_LEAVE_KERNEL(int32_t);
+REGISTER_REPEAT_INTER_LEAVE_KERNEL(int64_t);
```
Comment (zhongshsh) [inline oneflow/user/kernels/repeat_interleave_kernel.cpp:55]: in 如果是 float 类型也是支持的吧
 * (Tensor repeats, int dim, *, int output_size)
 * (int repeats, int dim, *, int output_size)

**Comment by zhongshsh** [comment]
如果 in 可以为 float 类型，那么需要补充求导逻辑

```python
>>> import torch 
>>> x = torch.tensor([1, 2, 3], dtype=torch.float32).requires_grad_()
>>> repeat = torch.tensor(2)
>>> y = x.repeat_interleave(repeat)
>>> y.sum().backward()
>>> x.grad
tensor([2., 2., 2.])
```

File: oneflow/core/functional/impl/array_functor.cpp [hunk 4]
```diff
@@ -2944,6 +2946,120 @@ class RepeatFunctor {
...
+  RepeatInterLeaveIntFunctor() {}
+  Maybe<Tensor> operator()(const std::shared_ptr<one::Tensor>& input, const int32_t& repeats,
+                           const Optional<int32_t>& dim) const {
+    CHECK_OR_RETURN(input->is_local() == true)
+        << Error::RuntimeError << "repeat_interleave only support local tensor now";
+    std::shared_ptr<one::Tensor> res;
+    if (!dim.has_value()) {
+      std::shared_ptr<one::Tensor> flatten_input = JUST(Flatten(input, 0, -1));
+      std::shared_ptr<one::Tensor> repeats_expand = JUST(
+          Expand(JUST(Constant(Shape{1}, Scalar(repeats), DType::Int32(), JUST(input->device()))),
```
[the remark this answers is not in the document]
Comment (Flowingsun007) [reply oneflow/core/functional/impl/array_functor.cpp:2982]: 好的

File: oneflow/core/functional/impl/array_functor.cpp [hunk 5]
```diff
@@ -2948,6 +2950,120 @@ class RepeatFunctor {
...
+    std::shared_ptr<one::Tensor> cumsum = JUST(Cumsum(repeats, 0, DType::Int32()));
+    int64_t output_size_value = 0;
+    for (const auto x : repeats_value) { output_size_value += x; }
+    std::shared_ptr<one::Tensor> res;
+    if (output_size_value > 0) {
+      res = JUST(IndexSelect(input, dim_,
+                             JUST(RepeatInterLeaveIndex(repeats, cumsum, output_size_value))));
+    } else {
+      DimVector new_input_shape(num_axes);
+      for (int i = 0; i < num_axes; i++) { new_input_shape[i] = input->shape()->At(i); }
```
Comment (Flowingsun007) [inline oneflow/core/functional/impl/array_functor.cpp:3056]: 这两行可以改成`DimVector new_input_shape(input_shape->dim_vec().begin(), input_shape->dim_vec().end());`

File: oneflow/core/functional/impl/array_functor.cpp [hunk 6]
```diff
@@ -2948,6 +2950,120 @@ class RepeatFunctor {
...
+    CHECK_OR_RETURN(repeats->dtype() == DType::Int64())
+        << Error::RuntimeError << "repeats has to be Long tensor";
+
+    std::vector<int64_t> repeats_value(repeats_shape->elem_cnt());
+    if (!output_size.has_value()) {
+      const auto& callback = [&](uint64_t ofblob_ptr) {
+        CHECK_JUST(BlobBufferCopyUtil<int64_t>::To(ofblob_ptr, repeats_value.data(),
+                                                   repeats_value.size()));
+      };
+      SyncAccessTensorWithTimeOut(repeats, callback, "const").GetOrThrow();
```
Comment (Flowingsun007) [inline oneflow/core/functional/impl/array_functor.cpp:3031]: 这个callback的操作，看上去主要是为了copy repeats的数值 -> repeats_value，然后后续对repeats_value做了check以及求和的操作，感觉都可以通过op来替换掉。
```python

# check
(repeats >=0).all()

# 求和
repeats.sum()
```

File: oneflow/core/functional/impl/array_functor.cpp [hunk 7]
```diff
@@ -2948,6 +2950,120 @@ class RepeatFunctor {
...
+    const int64_t num_axes = input_shape->NumAxes();
+    if (dim_ < 0) { dim_ += num_axes; }
+    CHECK_OR_RETURN(dim_ >= -num_axes && dim_ < num_axes)
+        << Error::IndexError() << "Dimension out of range (expected to be in range of ["
+        << -num_axes << ", " << num_axes - 1 << "], but got " << dim_ << ")";
+    CHECK_OR_RETURN(repeats_shape->At(0) == input->shape()->At(dim_))
+        << Error::RuntimeError << "repeats must have the same size as input along dim";
+    std::shared_ptr<one::Tensor> cumsum = JUST(Cumsum(repeats, 0, DType::Int32()));
+    int64_t output_size_value = 0;
+    for (const auto x : repeats_value) { output_size_value += x; }
```
Comment (Flowingsun007) [inline oneflow/core/functional/impl/array_functor.cpp:3049]: 建议改成
```c++
int64_t output_size_value = std::accumulate(repeats_value.begin(), repeats_value.end(), 0)
```

File: oneflow/core/functional/impl/array_functor.cpp (hunk 5 above)
Comment (BBuf) [reply oneflow/core/functional/impl/array_functor.cpp:3056]: 好的

File: oneflow/core/functional/impl/array_functor.cpp (hunk 7 above)
Comment (BBuf) [reply oneflow/core/functional/impl/array_functor.cpp:3049]: 好的

File: oneflow/user/kernels/repeat_interleave_kernel.cpp (hunk 3 above)
Comment (BBuf) [reply oneflow/user/kernels/repeat_interleave_kernel.cpp:55]: 不支持，这个输入tensor的语义是个数，只能用int表示

File: python/oneflow/framework/docstr/repeat_interleave.py (hunk 2 above)
Comment (BBuf) [reply python/oneflow/framework/docstr/repeat_interleave.py:24]: 好的

**Review by Flowingsun007 — APPROVED** [review]

**Review by zhongshsh — APPROVED** [review]

File: oneflow/core/functional/impl/array_functor.cpp [hunk 8]
```diff
@@ -2948,6 +2950,120 @@ class RepeatFunctor {
...
+          Expand(JUST(Constant(Shape{1}, Scalar(repeats), DType::Int32(), JUST(input->device()))),
+                 Shape{flatten_input->shape()->At(0)}));
+      std::shared_ptr<one::Tensor> cumsum = JUST(Cumsum(repeats_expand, 0, DType::Int32()));
+      int64_t output_size = flatten_input->shape()->At(0);
+      if (repeats > 0) { output_size *= repeats; }
+      res = JUST(IndexSelect(flatten_input, 0,
+                             JUST(RepeatInterLeaveIndex(repeats_expand, cumsum, output_size))));
+    } else {
+      int32_t dim_ = JUST(dim);
+      const auto input_shape = input->shape();
```
Comment (mosout) [inline oneflow/core/functional/impl/array_functor.cpp:2991]:
```suggestion
      const auto& input_shape = input->shape();
```

File: oneflow/core/functional/impl/array_functor.cpp [hunk 9]
```diff
@@ -2948,6 +2950,120 @@ class RepeatFunctor {
...
+class RepeatInterLeaveTensorFunctor {
+ public:
+  RepeatInterLeaveTensorFunctor() {}
+  Maybe<Tensor> operator()(const std::shared_ptr<one::Tensor>& input,
+                           const std::shared_ptr<one::Tensor>& repeats, const int32_t& dim,
+                           const Optional<int32_t>& output_size) const {
+    CHECK_OR_RETURN(input->is_local() == true)
+        << Error::RuntimeError() << "repeat_interleave only support local tensor now";
+    const auto repeats_shape = repeats->shape();
+    const int64_t& repeat_num_axes = repeats_shape->NumAxes();
```
Comment (mosout) [inline oneflow/core/functional/impl/array_functor.cpp:3019]:
```suggestion
    const auto& repeats_shape = repeats->shape();
    const int64_t repeat_num_axes = repeats_shape->NumAxes();
```

File: oneflow/core/functional/impl/array_functor.cpp [hunk 10]
```diff
@@ -2948,6 +2950,120 @@ class RepeatFunctor {
...
+      };
+      SyncAccessTensorWithTimeOut(repeats, callback, "const").GetOrThrow();
+      for (const auto x : repeats_value) {
+        CHECK_OR_RETURN(x >= 0) << Error::RuntimeError() << "repeats can not be negative";
+      }
+    } else {
+      repeats_value.push_back(JUST(output_size));
+    }
+    int32_t dim_ = dim;
+    const auto input_shape = input->shape();
```
Comment (mosout) [inline oneflow/core/functional/impl/array_functor.cpp:3039]:
```suggestion
    const auto& input_shape = input->shape();
```

**Review by mosout — APPROVED** [review]

File: oneflow/core/functional/impl/array_functor.cpp (hunk 8 above)
Comment (BBuf) [reply oneflow/core/functional/impl/array_functor.cpp:2991]: 好的
